crates/datasources/src/lib.rs

```diff
@@ -17,3 +30,49 @@ pub mod postgres;
 pub mod snowflake;
 pub mod sqlite;
 pub mod sqlserver;
+
+
+pub type DatasourceError = Box<dyn std::error::Error + Send + Sync>;
+#[async_trait]
+pub trait Datasource: Send + Sync {
+    fn name(&self) -> &'static str;
+
+    /// Validate the provided tunnel options.
+    fn validate_tunnel_connections(
+        &self,
+        tunnel_opts: Option<&TunnelOptions>,
+    ) -> Result<(), DatasourceError>;
+
+    /// Validate the provided credentials.
+    fn validate_credentials(
+        &self,
+        creds: Option<CredentialsOptions>,
+    ) -> Result<(), DatasourceError>;
+
+    /// Create a new datasource from the provided table options and credentials.
+    /// CREATE EXTERNAL TABLE foo FROM <name> OPTIONS (...) [CREDENTIALS] (...) [TUNNEL] (...)
+    // TODO: the datasource should have control over it's own CredentialsOptions and TunnelOptions
+    fn table_options_from_stmt(
+        &self,
+        opts: &mut StatementOptions,
```

Comment:
I don't think I understand why this needs to be mutable.

Reply:
we historically have been taking the values out of the statement via `.remove`. This avoids a few unnecessary `.clone()`.

It's important to keep in mind that this current `Datasource` definition is based on a lot of historical decisions such as using `Option<CredentialOptions>` and `Option<&TunnelOptions>`. The goal is just to first encapsulate everything into that trait, then we should be able to make incremental improvements much easier

Reply:
This makes sense: I'd like to have comments or TODOs for the things that call out the parts that are just cruft or legacy accommodations, because it's hard (as a reviewer) to see "what's the design" and "what's the implementation" in all of this.

Reply:
Why don't we just have this accept complete ownership? Is there a case where we want to use statement options after we create table options?